From 4e97690abedeb48640f07b9312c0fd3c1d7f3059 Mon Sep 17 00:00:00 2001
From: Karol Lewandowski <k.lewandowsk@samsung.com>
Date: Mon, 31 Dec 2012 14:35:36 +0100
Subject: [PATCH] Enable core dumps globally

diff --git a/src/system.conf b/src/system.conf
index 33d09bc..6c6b7f5 100644
--- a/src/system.conf
+++ b/src/system.conf
@@ -24,3 +24,4 @@
 #DefaultStandardOutput=journal
 #DefaultStandardError=inherit
 #JoinControllers=cpu,cpuacct
+DefaultLimitCORE=infinity
-- 
1.7.10.4

