From 69e7f67292ce52bf123958f392084ddae90a2ec1 Mon Sep 17 00:00:00 2001
From: Chris E Ferron <chris.e.ferron@linux.intel.com>
Date: Fri, 20 Jan 2012 11:27:56 -0800
Subject: [PATCH 2/2] systemd fsck disable l until linux

---
 src/fsck.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/src/fsck.c b/src/fsck.c
index c5088ad..7408654 100644
--- a/src/fsck.c
+++ b/src/fsck.c
@@ -322,7 +322,7 @@ int main(int argc, char *argv[]) {
         cmdline[i++] = "/sbin/fsck";
         cmdline[i++] = "-a";
         cmdline[i++] = "-T";
-        cmdline[i++] = "-l";
+        /* cmdline[i++] = "-l"; */ /* not supported in meego's util-linux */
 
         if (!root_directory)
                 cmdline[i++] = "-M";
-- 
1.7.7.5

