From 1391ba8699a8b96f80df63ae4be946dba51d404c Mon Sep 17 00:00:00 2001
From: Karol Lewandowski <k.lewandowsk@samsung.com>
Date: Fri, 13 Jul 2012 12:52:32 +0200
Subject: [PATCH] slp: Customize serial-getty@ unit for busybox's getty

---
 units/serial-getty@.service.m4 |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/units/serial-getty@.service.m4 b/units/serial-getty@.service.m4
index fc8b57b..9611e81 100644
--- a/units/serial-getty@.service.m4
+++ b/units/serial-getty@.service.m4
@@ -35,7 +35,7 @@ Before=getty.target
 
 [Service]
 Environment=TERM=vt100
-ExecStart=-/sbin/agetty -s %I 115200,38400,9600
+ExecStart=-/sbin/getty -L 115200 %I vt100
 Restart=always
 RestartSec=0
 UtmpIdentifier=%I
-- 
1.7.5.4

